modules/lang-painless/src/test/java/org/elasticsearch/painless/ElvisTests.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.painless;
+
+import static java.util.Collections.singletonMap;
+import static org.hamcrest.Matchers.greaterThan;
+import static org.hamcrest.Matchers.lessThan;
+
+/**
+ * Tests for the Elvis operator ({@code ?:}).
+ */
+public class ElvisTests extends ScriptTestCase {
+    public void testBasics() {
```

Comment:
The arguments for not having auto-boxing are the following:
1. Code maintenance -- we eliminate so many corner cases and potential bugs related to promotion and casting that this alone makes it worth not supporting this feature that will very rarely be used if ever.
2. Usability -- there is never a question of what to use, either a primitive type or def.  Again, the boxed types are whitelisted only to provide their utility methods which can be accessed directly from a primitive type or even a constant.  Also just because auto-boxing does not exist in a direct form, the user can still use the boxed type methods to move between a boxed and unboxed type if necessary (which I have trouble envisioning a valid use case).
3. Consistent casting behavior -- Painless has a very consistent casting model.  Groovy allows some crazy behavior such as allowing something like int to be assigned from a double ```def x = 2.0; int y = x;```, but then does not allow a Double to be passed into a method taking in an Integer.  The Java model does not work well with operators because Java only allows auto-boxing across similar types (int and Integer, double and Double), but is fine with promoting Integer to double for math operations because it unboxes first.  None of these things are great behavior.  Painless avoids all of this.  Promotion on the def type works exactly the same as promotion of primitives because if the def type is holding a boxed type it behaves as if it were a primitive.  If you pass in a def type that's a double into a method that takes an int, it won't be able to downcast.  If you try to do something like ```def x = 2.0; int y = x;``` It won't allow this to happen because it can't downcast.  There is no oddness of ```Integer y = 1; Double x = y;```  This works in Groovy, but not Java.  Painless just doesn't allow this at all and it avoids inconsistencies.

You can roll with your proposed model for these operators if you really want to, but they add little benefit given the extremely rare case that someone does use a field, and they will add some weird cases.  Consider this case with a FeatureTest object that I extended to have some public int fields --

```org.elasticsearch.painless.FeatureTest f = new org.elasticsearch.painless.FeatureTest(); f.x = 2; double x = f?.x ?: 1;```

This results in a ClassCastException because you cannot cast Integer to Double.  However, this is really strange because normally you can upcast int to double.  (And the casting model will never support the behavior of promoting boxed objects.)  This seems way more confusing to me than simply disallowing primitives for the current situation where the least obscure way to get a primitive type is to cast to a GeoPoint.  *If* we ever support compile-time field types then we can revisit this issue again.